### PR TITLE
core/config: check for stale Raft config

### PR DESCRIPTION
**Pull request by vickiniu**
See https://github.com/chain/chain/issues/1167

In https://github.com/chain/chain/pull/1385 and https://github.com/chain/chain/pull/1387, we write the config coreID to postgres whenever we configure a new cored process. This is to detect stale Raft configs resulting from when you restart a cored process, create a new Postgres db, but do not remove ~/.chaincore, preserving an old Raft config.

This PR checks the coreID stored in Raft against the coreID in Postgres (if any), and panics if there is a mismatch, suggesting that the user clean their Raft directory. Also adds a test to make sure we correctly detect stale configs.

### CODE REVIEW

File: core/config/config.go [hunk 1]
```diff
@@ -65,6 +65,10 @@ func Load(ctx context.Context, db pg.DB, sdb *sinkdb.DB) (*Config, error) {
 	if err != nil {
 		return nil, errors.Wrap(err)
 	} else if ver.Exists() {
+		raftID := c.Id
+		if !idMatchesPG(ctx, raftID, db) {
+			panic("Raft core ID doesn't match Postgres core ID: do you have a stale Raft config? Try: `rm -rf ~/.chaincore/raft`")
```
Comment (bobg) [inline core/config/config.go]: Don't hardcode `~/.chaincore/raft` here, populate it with the actual value, lest you advise people to remove the wrong directory (e.g. while testing with an alternate config).

Don't use backticks, which aren't meaningful in the console or log file where this message is likely to show up.

I might even consider rewording this message so that it doesn't contain a copy-and-paste command. For most things that would be a useful convenience, but for an `rm -rf` it's not a bad idea to require a little extra user intervention.

File: core/config/config.go [hunk 2]
```diff
@@ -99,6 +103,19 @@ func Load(ctx context.Context, db pg.DB, sdb *sinkdb.DB) (*Config, error) {
...
 }
 
+func idMatchesPG(ctx context.Context, id string, db pg.DB) bool {
+	const q = `SELECT id FROM core_id`
+	var pgID string
+	err := db.QueryRowContext(ctx, q).Scan(&pgID)
+	if err != nil {
+		return false
+	}
+	if pgID == id {
```
Comment (bobg) [inline core/config/config.go]: Could end the function here with `return pgID == id`

File: core/config/config.go (hunk 1 above)
Comment (kr) [inline core/config/config.go]: This should return an error, not panic.

File: core/config/config.go [hunk 3]
```diff
@@ -99,6 +103,19 @@ func Load(ctx context.Context, db pg.DB, sdb *sinkdb.DB) (*Config, error) {
...
+		return true
+	}
+	return false
```
Comment (kr) [inline core/config/config.go]: The last 7 lines of this function can (and perhaps should) be just one line.

File: core/config/config_test.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	}()
+
+	ctx := context.Background()
+	sdb := sinkdbtest.NewDB(t)
+	_, db := pgtest.NewDB(t, pgtest.SchemaPath)
+	c := newTestConfig(t)
+
+	// Write config to sinkdb
+	sdb.Exec(ctx,
+		sinkdb.IfNotExists("/core/config"),
```
Comment (kr) [inline core/config/config_test.go]: Nit: this line isn't really necessary since we know it's a fresh sinkdb, but nbd.

**Comment by vickiniu** [comment]
Unsure if we should handle these two cases in the same way:

- Raft has coreID, but Postgres has no coreID
- Raft has coreID, but Postgres has a different coreID

I'm not sure which scenario would land at the latter, but it seems stickier than just a stale Raft config. Right now, we throw the same error for both.

**Comment by kr** [comment]
Personally I think it's fine to treat those to cases the same, as long as the error message is phrased in a way that covers both.

File: core/config/env.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+import (
+	"os"
+	"path/filepath"
+)
+
+// HomeDirFromEnvironment returns the directory to use
+// for reading config and storing variable data.
+// It returns $CHAIN_CORE_HOME,
+// or, if that is empty, $HOME/.chaincore.
+func HomeDirFromEnvironment() string {
```
Comment (kr) [inline core/config/env.go]: It looks like there's no corresponding delete of this function from chain/core?

Also, maybe consider moving this function separately in its own PR first?

File: core/config/config.go [hunk 6]
```diff
@@ -65,6 +66,11 @@ func Load(ctx context.Context, db pg.DB, sdb *sinkdb.DB) (*Config, error) {
 	if err != nil {
 		return nil, errors.Wrap(err)
 	} else if ver.Exists() {
+		raftID := c.Id
+		if !idMatchesPG(ctx, raftID, db) {
+			raftDir := HomeDirFromEnvironment() + "/raft"
```
Comment (jbowens) [inline core/config/config.go]: might want to use [filepath.Join](https://godoc.org/path/filepath#Join) instead to ensure portability

File: core/config/config_test.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
+package config
+
+import (
+	"chain/database/pg/pgtest"
+	"chain/database/sinkdb"
+	"chain/database/sinkdb/sinkdbtest"
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"testing"
```
Comment (jbowens) [inline core/config/config_test.go]: nit: we group our imports with standard lib packages first, then vendored packages, then chain packages. like

```
import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"testing"

	"chain/database/pg/pgtest"
	"chain/database/sinkdb"
	"chain/database/sinkdb/sinkdbtest"
)
```

Comment (kr) [inline core/config/config_test.go]: If you're using `goimports` to automatically insert imports, you can use `$CHAIN/bin/gochainimports` as a drop-in replacement that will do this grouping automatically.

File: core/config/env.go (hunk 5 above)
Comment (vickiniu) [inline core/config/env.go]: Thanks for the suggestion! :) https://github.com/chain/chain/pull/1392

**Comment by vickiniu** [comment]
Looks like the diff is still showing changes against `main` from when I opened the PR, not including recent commits on `main`. 

I rebased this branch to include moving `env.go` from core -> config (https://github.com/chain/chain/pull/1392#issuecomment-312107190) but those changes are all showing up in the diff. Not sure if there's a good way to fix this?

File: core/config/config.go [hunk 8]
```diff
@@ -99,6 +106,13 @@ func Load(ctx context.Context, db pg.DB, sdb *sinkdb.DB) (*Config, error) {
 	return c, nil
 }
 
+func idMatchesPG(ctx context.Context, id string, db pg.DB) bool {
+	const q = `SELECT id FROM core_id`
+	var pgID string
+	err := db.QueryRowContext(ctx, q).Scan(&pgID)
+	return err == nil && pgID == id
+}
```
Comment (jbowens) [inline core/config/config.go]: If there's an error querying the database here, we'll end up returning ErrStaleRaftConfig from Load which is potentially inaccurate. Maybe we should add an error return value to differentiate between trouble connecting to the database and a wiped Postgres database.

Heads up that If there is no Core ID in the table, QueryRowContext will return [sql.ErrNoRows](https://godoc.org/database/sql#pkg-variables). idMatchesPG could translate that specific error into a `(false, nil)` return value and return any other error as-is.

**Comment by jbowens** [comment]
looks great!

File: core/config/config.go [hunk 9]
```diff
@@ -99,6 +109,16 @@ func Load(ctx context.Context, db pg.DB, sdb *sinkdb.DB) (*Config, error) {
 	return c, nil
 }
 
+func idMatchesPG(ctx context.Context, id string, db pg.DB) (bool, error) {
+	const q = `SELECT id FROM core_id`
+	var pgID string
+	err := db.QueryRowContext(ctx, q).Scan(&pgID)
+	if err != sql.ErrNoRows {
```
Comment (kr) [inline core/config/config.go]: If we get any other error here, shouldn't we treat that as an error, rather than "not a match"? I would have expected

```go
if err != nil {
```

here

**Comment by kr** [comment]
LGTM
